types/lodash/common/object.d.ts

```diff
@@ -3111,9 +3111,9 @@ declare module "../index" {
          * _.pickBy(object, _.isNumber);
          * // => { 'a': 1, 'c': 3 }
          */
-        pickBy<T extends object, S extends T[keyof T]>(
-            object: T | null | undefined,
-            predicate: ObjectIteratorTypeGuard<T, S>
+        pickBy<T, S extends T>(
+            object: Dictionary<T> | null | undefined,
```

Comment:
Actually, it looks like the non-type-guarded version of `pickBy` already has 3 overloads (`Dictionary`, `NumericDictionary`, `object`). So I think we should mirror that in the type guarded version. So there should be 6 overloads total (3 type-guarded, 3 non-type-guarded)

Reply:
``` ts
    interface MixedRecordInterface {
        a: string;
        b: number;
    }
    const mixedRecordInterface: MixedRecordInterface = anything;

    _.pickBy(mixedRecordInterface, (item: string | number): item is number => typeof item === "number"); // $ExpectType Partial<MixedRecord>

    type MixedRecordType = {
        a: string;
        b: number;
    }
    const mixedRecordType: MixedRecordType = anything;

    _.pickBy(mixedRecordType, (item: string | number): item is number => typeof item === "number"); // $ExpectType Dictionary<number>
```

These tests already passes, thanks to these overloads:

``` ts
        pickBy<T>(
            object: Dictionary<T> | null | undefined,
            predicate?: ValueKeyIteratee<T>
        ): Dictionary<T>;

        pickBy<T extends object>(
            object: T | null | undefined,
            predicate?: ValueKeyIteratee<T[keyof T]>
        ): PartialObject<T>;
```

We don't need a type guard specific overload for the record object, since we can't narrow the record type to just the filtered keys (e.g. `{ b: number }`) anyway. (AFAIK that's not possible in TS.)

(If we pass a type alias instead of an interface, the index signature overload is used instead because of https://github.com/Microsoft/TypeScript/issues/15300.)

Thus, I think we only need the overload for a predicate with a user-defined type guard when the object passed in is a dictionary, not a record.

I will add an overload for `NumericDirectionary`.

Reply:
After adding overloads for `NumericDictionary`, I now have a different problem.

``` ts
/* Argument of type 'NumericDictionary<string | number> | null | undefined' is not assignable to parameter of type 'Dictionary<string | number> | null | undefined'.
        Type 'NumericDictionary<string | number>' is not assignable to type 'Dictionary<string | number>'.
            Index signature is missing in type 'NumericDictionary<string | number>'. */
    fp.pickBy((item: string | number): item is number => typeof item === "number")(mixedNumericDictionary); // $ExpectType NumericDictionary<number>
```

This is because the same it's using the `Dictionary` overload instead of the `NumericDictionary` overload, as `Dictionary` comes first and TS has no way of knowing we're going to pass a `NumericDictionary` later on… 😒

Reply:
Currently, `fp.pickBy` has overloads for both `Dictionary` and `NumericDictionary`.

However, this isn't practical because we don't pass the dictionary to this function, but instead to the resulting function.

I guess we need the call to `fp.pickBy()` to return a function which has two overloads, one for `Dictionary` and one for `NumericDictionary`.

Is this possible?

/cc @aj-r (Thank you for all your help on this!)

Reply:
> I guess we need the call to fp.pickBy() to return a function which has two overloads, one for Dictionary and one for NumericDictionary.

The fp generator should do that automatically as long as the type of the second parameter (`predicate`) is identical for both overloads. I think you could use `ObjectIteratorTypeGuard<Dictionary<T>` for both overloads, although the third callback parameter won't quite have the right type.

However that brings up an interesting point - the callback for `pickBy` has 2 parameters, but `ObjectIteratorTypeGuard` has 3 parameters. So maybe `ObjectIteratorTypeGuard` isn't the right type to use at all. Maybe you need to make a new callback type (`ValueKeyIteratorTypeGuard`?) and use that for both the `Dictionary` and `NumericDictionary` overloads.

Reply:
Thanks, I've fixed that now. I awkwardly called it `ObjectIteratorTypeGuardWithoutCollection`. Any better ideas?

I wonder why `pickBy` doesn't receive the `collection` param… 🤔